File: clickbait_py/sources/hooktheory.py
"""Hooktheory TheoryTab — key and song section data via page scraping.

No API key required. Scrapes the public TheoryTab page for key signatures
and section structure. Coverage depends on community contributions (~72k songs).
"""

import re

import requests
from bs4 import BeautifulSoup


THEORYTAB_BASE = "https://www.hooktheory.com/theorytab/view"


def _slugify(text: str) -> str:
    """Convert text to URL slug format matching TheoryTab conventions."""
    return re.sub(r"[^a-z0-9]+", "-", text.lower()).strip("-")


def lookup_song(title: str, artist: str) -> dict | None:
    """Look up a song on TheoryTab for key and section data.

    Requires both title and artist to construct the URL slug.
    Returns None if the song isn't in the database.
    """
    artist_slug = _slugify(artist)
    song_slug = _slugify(title)
    url = f"{THEORYTAB_BASE}/{artist_slug}/{song_slug}"

    resp = requests.get(url, timeout=10)
    if resp.status_code != 200:
        return None

    soup = BeautifulSoup(resp.text, "html.parser")

    # Extract keys
    keys = []
    key_text = soup.find(string=re.compile("analyzed in the following keys"))
    if key_text:
        parent = key_text.find_parent("p")
        if parent:
            keys = [
                a.text.strip()
                for a in parent.find_all("a")
                if "cheat-sheet/key" in a.get("href", "")
            ]

    # Extract sections from anchor links to #section-name
    section_links = soup.find_all("a", href=re.compile(f"{song_slug}#"))
    seen = set()
    sections = []
    for link in section_links:
        span = link.find("span")
        if span:
            name = span.text.strip()
            if name and name not in seen:
                sections.append(name)
                seen.add(name)

    if not keys and not sections:
        return None

    return {
        "title": title,
        "artist": artist,
        "keys": keys,
        "sections": sections,
        "url": url,
    }

File: crates/audio/src/lookup/hooktheory.rs
use anyhow::Result;
use scraper::{Html, Selector};

const THEORYTAB_BASE: &str = "https://www.hooktheory.com/theorytab/view";

fn slugify(text: &str) -> String {
    let mut slug = String::new();
    let mut last_was_dash = false;
    for c in text.chars() {
        if c.is_ascii_alphanumeric() {
            slug.push(c.to_ascii_lowercase());
            last_was_dash = false;
        } else if !last_was_dash && !slug.is_empty() {
            slug.push('-');
            last_was_dash = true;
        }
    }
    slug.trim_end_matches('-').to_string()
}

#[derive(Debug)]
pub struct HooktheoryResult {
    pub title: String,
    pub artist: String,
    pub keys: Vec<String>,
    pub sections: Vec<String>,
    pub url: String,
}

pub async fn lookup_song(title: &str, artist: &str) -> Result<Option<HooktheoryResult>> {
    let artist_slug = slugify(artist);
    let song_slug = slugify(title);
    let url = format!("{}/{}/{}", THEORYTAB_BASE, artist_slug, song_slug);

    let resp = reqwest::get(&url).await?;
    if !resp.status().is_success() {
        return Ok(None);
    }

    let html = resp.text().await?;
    let doc = Html::parse_document(&html);

    // Extract keys
    let mut keys = Vec::new();
    let p_selector = Selector::parse("p").unwrap();
    for p in doc.select(&p_selector) {
        let text = p.text().collect::<String>();
        if text.contains("analyzed in the following keys") {
            let a_selector = Selector::parse("a").unwrap();
            for a in p.select(&a_selector) {
                if let Some(href) = a.value().attr("href") {
                    if href.contains("cheat-sheet/key") {
                        let key_text: String = a.text().collect();
                        let trimmed = key_text.trim().to_string();
                        if !trimmed.is_empty() {
                            keys.push(trimmed);
                        }
                    }
                }
            }
        }
    }

    // Extract sections from anchor links
    let section_pattern = format!("{}#", song_slug);
    let a_selector = Selector::parse("a").unwrap();
    let span_selector = Selector::parse("span").unwrap();
    let mut sections = Vec::new();
    let mut seen = std::collections::HashSet::new();

    for a in doc.select(&a_selector) {
        if let Some(href) = a.value().attr("href") {
            if href.contains(&section_pattern) {
                for span in a.select(&span_selector) {
                    let name: String = span.text().collect();
                    let name = name.trim().to_string();
                    if !name.is_empty() && seen.insert(name.clone()) {
                        sections.push(name);
                    }
                }
            }
        }
    }

    if keys.is_empty() && sections.is_empty() {
        return Ok(None);
    }

    Ok(Some(HooktheoryResult {
        title: title.to_string(),
        artist: artist.to_string(),
        keys,
        sections,
        url,
    }))
}
